README.md
# rio-chop

<p align="center">
  <em>🔪🔪🔪 Split a dataset into smaller COGs</em>
</p>
<p align="center">
  <a href="https://github.com/cogeotiff/rio-chop/actions?query=workflow%3ACI" target="_blank">
      <img src="https://github.com/cogeotiff/rio-chop/workflows/CI/badge.svg" alt="Test">
  </a>
  <a href="https://codecov.io/gh/cogeotiff/rio-chop" target="_blank">
      <img src="https://codecov.io/gh/cogeotiff/rio-chop/branch/master/graph/badge.svg" alt="Coverage">
  </a>
  <a href="https://pypi.org/project/rio-chop" target="_blank">
      <img src="https://img.shields.io/pypi/v/rio-chop?color=%2334D058&label=pypi%20package" alt="Package version">
  </a>
  <a href="https://pypistats.org/packages/rio-chop" target="_blank">
      <img src="https://img.shields.io/pypi/dm/rio-chop.svg" alt="Downloads">
  </a>
  <a href="https://github.com/cogeotiff/rio-chop/blob/master/LICENSE" target="_blank">
      <img src="https://img.shields.io/github/license/cogeotiff/rio-chop.svg" alt="Downloads">
  </a>
</p>

---

**Source Code**: <a href="https://github.com/cogeotiff/rio-chop" target="_blank">https://github.com/cogeotiff/rio-chop</a>

---

`rio-chop` is a simple [rasterio](https://github.com/rasterio/rasterio) plugin for creating smaller COGs from an input dataset.

## Installation

```bash
$ pip install pip -U

# From Pypi
$ pip install rio-chop

# Or from source
$ pip install git+http://github.com/cogeotiff/rio-chop
```

## GDAL >3.1

`rio-chop` is using the new `COG` driver, thus GDAL >=3.1 is required.

## Usage
```
$ rio chop --help
Usage: rio chop [OPTIONS] INPUT

  Split a dataset into smaller COGs

Options:
  --prefix TEXT                   output file prefix.  [required]
  -w, --width INTEGER             Chop width (default: 5000px)
  -h, --height INTEGER            Chop height (default: 5000px)
  -p, --cog-profile [jpeg|webp|zstd|lzw|deflate|packbits|lzma|lerc|lerc_deflate|lerc_zstd|raw]
                                  CloudOptimized GeoTIFF profile (default:
                                  deflate).

  --blocksize INTEGER             Overwrite profile's tile size.
  --overview-level INTEGER        Overview level (if not provided, appropriate
                                  overview level will be selected until the
                                  smallest overview is smaller than the value
                                  of the internal blocksize)

  --overview-resampling [nearest|bilinear|cubic|cubic_spline|lanczos|average|mode|gauss|rms]
                                  Overview creation resampling algorithm
                                  (default: nearest).

  --overview-blocksize TEXT       Overview's internal tile size (default
                                  defined by GDAL_TIFF_OVR_BLOCKSIZE env or
                                  128)

  -w, --web-optimized             Create COGEO optimized for Web.
  --zoom-level-strategy [lower|upper|auto]
                                  Strategy to determine zoom level. (default:
                                  auto).

  --aligned-levels INTEGER        Number of overview levels for which GeoTIFF
                                  tile and tiles defined in the tiling scheme
                                  match.

  -r, --resampling [nearest|bilinear|cubic|cubic_spline|lanczos|average|mode|max|min|med|q1|q3|sum|rms]
                                  Resampling algorithm (default: nearest).
                                  Will only be applied with the `--web-
                                  optimized` option.

  --co, --profile NAME=VALUE      Driver specific creation options. See the
                                  documentation for the selected output driver
                                  for more information.

  --config NAME=VALUE             GDAL configuration options.
  --help                          Show this message and exit.
```

### Examples

```
$ rio chop world_cog.tif --prefix "world" -p jpeg --blocksize 256
  [####################################]  100%

$ ls -1 world*
world_0_0.tif
world_0_10000.tif
world_0_5000.tif
world_10000_0.tif
world_10000_10000.tif
world_10000_5000.tif
world_15000_0.tif
world_15000_10000.tif
world_15000_5000.tif
world_20000_0.tif
world_20000_10000.tif
world_20000_5000.tif
world_5000_0.tif
world_5000_10000.tif
world_5000_5000.tif
world_cog.tif
```

## Contribution & Development

See [CONTRIBUTING.md](https://github.com/cogeotiff/rio-chop/blob/main/CONTRIBUTING.md)

## Changes

See [CHANGES.md](https://github.com/cogeotiff/rio-chop/blob/main/CHANGES.md).

## License

See [LICENSE](https://github.com/cogeotiff/rio-chop/blob/main/LICENSE)



CHANGES.md
# Release Notes

# 0.1.0

* Initial release.


tests/conftest.py
"""``pytest`` configuration."""

import pytest


@pytest.fixture
def runner():
    """CLI Runner fixture."""
    from click.testing import CliRunner

    return CliRunner()


tests/test_cli.py
"""tests rio_chop.cli."""

import os

import rasterio
from rio_cogeo import cog_validate, utils

from rio_chop.scripts.cli import chop

PREFIX = os.path.join(os.path.dirname(__file__), "fixtures")


def test_rio_chop_cli(runner):
    """Should work as expected."""
    with runner.isolated_filesystem():
        src_path = os.path.join(PREFIX, "image.tif")
        result = runner.invoke(
            chop,
            [
                src_path,
                "--prefix",
                "test",
                "--width",
                "500",
                "--height",
                "500",
                "--blocksize",
                "256",
            ],
        )
        assert not result.exception
        assert result.exit_code == 0
        assert os.path.exists("test_0_0.tif")
        assert os.path.exists("test_0_500.tif")
        assert os.path.exists("test_500_0.tif")
        assert os.path.exists("test_500_500.tif")

        # make sure we create valid COG
        assert cog_validate("test_500_500.tif")

    # Colormap
    with runner.isolated_filesystem():
        src_path = os.path.join(PREFIX, "image_colormap.tif")
        result = runner.invoke(
            chop,
            [
                src_path,
                "--prefix",
                "test_colormap",
                "--width",
                "500",
                "--height",
                "500",
                "--blocksize",
                "256",
            ],
        )
        assert not result.exception
        assert result.exit_code == 0
        assert os.path.exists("test_colormap_0_0.tif")
        assert os.path.exists("test_colormap_0_500.tif")
        assert os.path.exists("test_colormap_500_0.tif")
        assert os.path.exists("test_colormap_500_500.tif")

        # make sure we create valid COG
        assert cog_validate("test_colormap_500_500.tif")

        with rasterio.open(src_path) as src, rasterio.open(
            "test_colormap_500_500.tif"
        ) as cog:
            assert cog.colormap(1) == src.colormap(1)
            assert cog.crs == src.crs

    # Tags
    with runner.isolated_filesystem():
        src_path = os.path.join(PREFIX, "image_tags.tif")
        result = runner.invoke(
            chop,
            [
                src_path,
                "--prefix",
                "test_tags",
                "--width",
                "500",
                "--height",
                "500",
                "--blocksize",
                "256",
            ],
        )
        assert not result.exception
        assert result.exit_code == 0
        assert os.path.exists("test_tags_0_0.tif")
        assert os.path.exists("test_tags_0_500.tif")
        assert os.path.exists("test_tags_500_0.tif")
        assert os.path.exists("test_tags_500_500.tif")

        # make sure we create valid COG
        assert cog_validate("test_tags_0_0.tif")

        with rasterio.open(src_path) as src, rasterio.open("test_tags_0_0.tif") as cog:
            assert cog.tags()["DatasetName"] == cog.tags()["DatasetName"]

    with runner.isolated_filesystem():
        src_path = os.path.join(PREFIX, "image.tif")
        result = runner.invoke(
            chop,
            [
                src_path,
                "--prefix",
                "test_web",
                "--width",
                "500",
                "--height",
                "500",
                "--blocksize",
                "256",
                "-w",
            ],
        )
        assert not result.exception
        assert result.exit_code == 0
        assert os.path.exists("test_web_0_0.tif")
        assert os.path.exists("test_web_0_500.tif")
        assert os.path.exists("test_web_500_0.tif")
        assert os.path.exists("test_web_500_500.tif")

        # make sure we create valid COG
        assert cog_validate("test_web_500_500.tif")

        with rasterio.open("test_web_500_500.tif") as cog:
            assert cog.crs.to_epsg() == 3857

    with runner.isolated_filesystem():
        src_path = os.path.join(PREFIX, "image_rgb_mask.tif")
        result = runner.invoke(
            chop,
            [
                src_path,
                "--prefix",
                "test_mask",
                "--width",
                "500",
                "--height",
                "500",
                "--blocksize",
                "256",
                "-p",
                "jpeg",
            ],
        )
        assert not result.exception
        assert result.exit_code == 0
        assert os.path.exists("test_mask_0_0.tif")
        assert os.path.exists("test_mask_0_500.tif")
        assert os.path.exists("test_mask_500_0.tif")
        assert os.path.exists("test_mask_500_500.tif")

        # make sure we create valid COG
        assert cog_validate("test_mask_500_500.tif")

        with rasterio.open("test_mask_500_500.tif") as cog:
            assert utils.has_mask_band(cog)
            assert cog.profile["compress"] == "jpeg"


rio_chop/scripts/cli.py
"""Chop."""

import math
import os
import sys
import warnings
from contextlib import ExitStack

import click
import morecantile
import rasterio
from rasterio.dtypes import _gdal_typename
from rasterio.enums import ColorInterp
from rasterio.enums import Resampling as ResamplingEnums
from rasterio.env import GDALVersion
from rasterio.rio import options
from rasterio.shutil import copy
from rasterio.warp import SUPPORTED_RESAMPLING as WarpResampling
from rasterio.windows import Window
from rio_cogeo import cog_profiles, utils

OverviewResampling = [r for r in ResamplingEnums if r.value < 8]
if GDALVersion.runtime().at_least("3.3"):
    OverviewResampling.append(ResamplingEnums.rms)


def dims(total, chop):
    """Given a total number of pixels, chop into equal chunks.

    Last one gets truncated so that sum of sizes == total.
    yeilds (offset, size) tuples
    >>> list(dims(512, 256))
    [(0, 256), (256, 256)]
    >>> list(dims(502, 256))
    [(0, 256), (256, 246)]
    >>> list(dims(522, 256))
    [(0, 256), (256, 256), (512, 10)]
    """
    for a in range(int(math.ceil(total / chop))):
        offset = a * chop
        diff = total - offset
        if diff <= chop:
            size = diff
        else:
            size = chop
        yield offset, size


@click.command()
@options.file_in_arg
@click.option(
    "--prefix", type=str, required=True, help="output file prefix.",
)
@click.option(
    "--width", type=int, default=5000, help="Chop width (default: 5000px)",
)
@click.option("--height", type=int, default=5000, help="Chop height (default: 5000px)")
@click.option(
    "--cog-profile",
    "-p",
    "cogeo_profile",
    type=click.Choice(cog_profiles.keys(), case_sensitive=False),
    default="deflate",
    help="CloudOptimized GeoTIFF profile (default: deflate).",
)
@click.option("--blocksize", type=int, help="Overwrite profile's tile size.")
@click.option(
    "--overview-level",
    type=int,
    help="Overview level (if not provided, appropriate overview level will be "
    "selected until the smallest overview is smaller than the value of the "
    "internal blocksize)",
)
@click.option(
    "--overview-resampling",
    help="Overview creation resampling algorithm (default: nearest).",
    type=click.Choice([it.name for it in OverviewResampling]),
    default="nearest",
)
@click.option(
    "--overview-blocksize",
    default=lambda: os.environ.get("GDAL_TIFF_OVR_BLOCKSIZE", 128),
    help="Overview's internal tile size (default defined by "
    "GDAL_TIFF_OVR_BLOCKSIZE env or 128)",
)
@click.option(
    "--web-optimized", "-w", is_flag=True, help="Create COGEO optimized for Web."
)
@click.option(
    "--zoom-level-strategy",
    type=click.Choice(["lower", "upper", "auto"], case_sensitive=False),
    default="auto",
    help="Strategy to determine zoom level. (default: auto).",
)
@click.option(
    "--aligned-levels",
    type=int,
    help="Number of overview levels for which GeoTIFF tile and tiles defined in the tiling scheme match.",
)
@click.option(
    "--resampling",
    "-r",
    help="Resampling algorithm (default: nearest). Will only be applied with the `--web-optimized` option.",
    type=click.Choice([it.name for it in WarpResampling]),
    default="nearest",
)
@options.creation_options
@click.option(
    "--config",
    "config",
    metavar="NAME=VALUE",
    multiple=True,
    callback=options._cb_key_val,
    help="GDAL configuration options.",
)
@click.option(
    "--quiet", "-q", help="Remove progressbar and other non-error output.", is_flag=True
)
def chop(
    input,
    prefix,
    width,
    height,
    cogeo_profile,
    blocksize,
    overview_level,
    overview_resampling,
    overview_blocksize,
    web_optimized,
    zoom_level_strategy,
    aligned_levels,
    resampling,
    creation_options,
    config,
    quiet,
):
    """Split a dataset into smaller COGs"""
    if not GDALVersion.runtime().at_least("3.1"):
        raise Exception(
            f"Incompatible GDAL version {GDALVersion.runtime()}. Must be at least 3.1."
        )

    output_profile = cog_profiles.get(cogeo_profile)
    output_profile.update(dict(BIGTIFF=os.environ.get("BIGTIFF", "IF_SAFER")))
    if creation_options:
        output_profile.update(creation_options)

    if blocksize:
        output_profile["blockxsize"] = blocksize
        output_profile["blockysize"] = blocksize

    if web_optimized:
        overview_blocksize = blocksize or 512

    config.update(
        dict(
            GDAL_NUM_THREADS="ALL_CPUS",
            GDAL_TIFF_INTERNAL_MASK=os.environ.get("GDAL_TIFF_INTERNAL_MASK", True),
            GDAL_TIFF_OVR_BLOCKSIZE=str(overview_blocksize),
        )
    )

    tms = morecantile.tms.get("WebMercatorQuad")

    tilesize = min(int(output_profile["blockxsize"]), int(output_profile["blockysize"]))

    with rasterio.Env(**config):
        with ExitStack() as ctx:
            src = ctx.enter_context(rasterio.open(input))

            w, h = src.meta["width"], src.meta["height"]
            mask = utils.has_mask_band(src)

            winds = [
                (coff, wd, roff, ht)
                for roff, ht in dims(h, height)
                for coff, wd in dims(w, width)
            ]

            fout = ctx.enter_context(open(os.devnull, "w")) if quiet else sys.stderr
            with click.progressbar(
                winds, length=len(winds), file=fout, show_percent=True
            ) as blocks:
                for col_off, w, row_off, h in blocks:

                    window = Window(col_off, row_off, w, h)
                    data = src.read(window=window)

                    # We are using GDAL MEM driver to create a new dataset from the numpy array
                    # ref: https://github.com/rasterio/rasterio/blob/master/rasterio/_io.pyx#L1946-L1955
                    info = {
                        "DATAPOINTER": data.__array_interface__["data"][0],
                        "PIXELS": w,
                        "LINES": h,
                        "BANDS": src.count,
                        "DATATYPE": _gdal_typename(data.dtype.name),
                    }
                    dataset_options = ",".join(
                        f"{name}={val}" for name, val in info.items()
                    )
                    datasetname = f"MEM:::{dataset_options}"

                    with rasterio.open(datasetname, "r+") as w_dst:
                        # Set transform, CRS and nodata
                        w_dst.transform = src.window_transform(window)
                        w_dst.crs = src.crs
                        w_dst.nodata = src.nodata
                        w_dst.colorinterp = src.colorinterp

                        # Forward colormap
                        if src.colorinterp[0] is ColorInterp.palette:
                            try:
                                w_dst.write_colormap(1, src.colormap(1))
                            except ValueError:
                                pass

                        if mask:
                            w_dst.write_mask(
                                src.dataset_mask(window=window).astype("uint8")
                            )

                        # Forward Tags/Descriptions
                        w_dst.update_tags(**src.tags())
                        w_dst._set_all_scales([src.scales[b - 1] for b in src.indexes])
                        w_dst._set_all_offsets(
                            [src.offsets[b - 1] for b in src.indexes]
                        )

                        for i, b in enumerate(src.indexes):
                            w_dst.set_band_description(i + 1, src.descriptions[b - 1])
                            w_dst.update_tags(i + 1, **src.tags(b))

                        output_profile["driver"] = "COG"  # COG Driver
                        if web_optimized:
                            output_profile["TILING_SCHEME"] = (
                                "GoogleMapsCompatible"
                                if tms.identifier == "WebMercatorQuad"
                                else tms.identifier
                            )

                            if mask and output_profile.get("compress", "") != "JPEG":
                                warnings.warn(
                                    "With GDAL COG driver, mask band will be translated to an alpha band."
                                )

                        output_profile["zoom_level_strategy"] = zoom_level_strategy
                        output_profile["overview_resampling"] = overview_resampling
                        output_profile["warp_resampling"] = resampling
                        if aligned_levels is not None:
                            output_profile["aligned_levels"] = aligned_levels

                        output_profile["blocksize"] = tilesize
                        output_profile.pop("blockxsize", None)
                        output_profile.pop("blockysize", None)
                        output_profile.pop("tiled", None)
                        output_profile.pop("interleave", None)
                        output_profile.pop("photometric", None)

                        copy(
                            w_dst, f"{prefix}_{col_off}_{row_off}.tif", **output_profile
                        )
